dokka-runners/dokka-gradle-plugin/src/main/kotlin/internal/v2MigrationUtils.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package org.jetbrains.dokka.gradle.internal
+
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.ConfigurationContainer
+import org.gradle.internal.deprecation.DeprecatableConfiguration
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.withType
+import org.jetbrains.dokka.gradle.DokkaCollectorTask
+import org.jetbrains.dokka.gradle.DokkaMultiModuleTask
+import org.jetbrains.dokka.gradle.DokkaTask
+import org.jetbrains.dokka.gradle.DokkaTaskPartial
+import org.jetbrains.dokka.gradle.dependencies.DependencyContainerNames
+
+/**
+ * Add some utilities to make migrating from V1 to V2 easier.
+ *
+ * - Create the same tasks as V1, but disable them and update the description.
+ * - Create the same Configurations as V1, if they are missing from V2, and mark them as deprecated.
+ */
+internal fun addV2MigrationHelpers(
+    project: Project
+) {
+    // Note: `dokkaPlugin` exists in the v2 plugin, so no migration helper is required.
+    project.configurations.createDokkaDefaultRuntimeConfiguration()
+
+    setupDokkaTasks(project, "GFM")
+    setupDokkaTasks(project, "Javadoc", multiModuleTaskSupported = false)
+    setupDokkaTasks(project, "Jekyll")
+    setupDokkaTasks(project, "HTML", createDokkaPluginFormatConfiguration = false)
+
+    configureDokkaTaskConventions(project)
+}
+
+private fun configureDokkaTaskConventions(project: Project) {
+    project.tasks.withType<DokkaTask>().configureEach {
+        group = "" // hide from users
+        onlyIf("Dokka V1 tasks are disabled due to the V2 flag being enabled") { false }
```

Comment:
Hmmm maybe, although I would prefer to be lenient here. The errors will come when people disable the migration helper flags!

We could add a `doFirst { logger.warn("this task does nothing, check $migrationGuideLink") }` though, to help people discover any issues?

Reply:
the problem is that the task names are different per V1 and V2 and so I would expect that:
* user will enable v2 (migration helpers should be enabled automatically)
* probably in most cases no changes in build script will be required
* user will run `dokkaHtml` - task (and the whole build) will be executed successfully (even if it will show `dokkaHtml` as `skipped`)
Done. User migrated. But do they?...

Logging a warning is better, but still, the build will be successful while doing literally nothing...

Is there any reason to not fail this task which is needed only for migration? Especially, when we will flip the flag later.

Reply:
I see the progression as being:

1. User updates Dokka and enables the v2 plugin using `enableV2Migration`.
2. They finish migration, but accidentally leave references to old Dokka tasks in the build scripts.
3. In order to suppress the warning, they replace `enableV2Migration=true` with `enableV2=true`.
4. Boom, the build script explodes.

Reply:
Yeah, I see your point. But I don't see how `skipping` this task is more `lenient`.
TBH, I don't really have a strong preference here. Also this could be adapted later if needed.
But at least some warning should be there, silent skipping of the main task from old DGP is not nice

Reply:
For example, let's say a project uses Dokka, but also uses an external plugin that requires the Dokka tasks are present and can run (or can be skipped). The external plugin will need to be updated eventually, but in the meantime, the migration helpers will at least let it work.

Gradle buildscripts can be _really_ weird, so avoiding throwing errors is best.

We can create a 'warning: Dokka tasks are disabled' log though. It'd be better to do it in a different PR though.

Reply:
Good point regarding external plugins! Thanks!

Reply:
I think I agree here with Adam - we should fail the build in this case and failures will happen when user will switch off these helpers.

Reply:
> we should fail the build in this case

Did you mean "we shouldn't fail the build in this case (when migration helpers are enabled)"?

Reply:
Sorry, yes - we should not fail the build when migration helpers are enabled.